lib-rt/CPy.h

```diff
@@ -69,24 +70,117 @@ static inline void CPy_FixupTraitVtable(CPyVTableItem *vtable, int count) {
 
 // Create a heap type based on a template non-heap type.
 // This is super hacky and maybe we should suck it up and use PyType_FromSpec instead.
-static inline PyTypeObject *CPyType_FromTemplate(PyTypeObject *template_,
-                                                 PyObject *modname) {
-    PyHeapTypeObject *t = (PyHeapTypeObject*)PyType_GenericAlloc(&PyType_Type, 0);
+static inline PyObject *CPyType_FromTemplate(PyTypeObject *template_,
+                                             PyObject *bases,
+                                             PyObject *modname) {
+    PyHeapTypeObject *t = NULL;
+    PyTypeObject *dummy_class = NULL;
+    PyObject *name;
+
+    PyTypeObject *metaclass = Py_TYPE(template_);
+
+    PyObject *old_bases = bases;
+    if (bases) {
+        Py_INCREF(old_bases);
+        bases = update_bases(old_bases);
+
+        // Find the appropriate metaclass from our base classes. We
+        // care about this because Generic uses a metaclass prior to
+        // Python 3.7.
+        metaclass = _PyType_CalculateMetaclass(metaclass, bases);
+        if (!metaclass)
+            goto error;
+    }
+
+    name = PyUnicode_FromString(template_->tp_name);
+    if (!name)
+        goto error;
+
+    // If there is a metaclass other than type, we would like to call
+    // its __new__ function. Unfortunately there doesn't seem to be a
+    // good way to mix a C extension class and creating it via a
+    // metaclass. We need to do it anyways, though, in order to
+    // support subclassing Generic[T] prior to Python 3.7.
+    //
+    // We solve this with a kind of atrocious hack: create a parallel
+    // class using the metaclass, determine the bases of the real
+    // class by pulling them out of the parallel class, creating the
+    // real class, and then merging its dict back into the original
+    // class. There are lots of cases where this won't really work,
+    // but for the case of GenericMeta setting a bunch of properties
+    // on the class we should be fine.
+    if (metaclass != &PyType_Type) {
+        PyObject *ns = PyDict_New();
+        if (!ns) goto error;
+
+        dummy_class = (PyTypeObject *)PyObject_CallFunctionObjArgs(
+            (PyObject *)metaclass, name, bases, ns, NULL);
+        Py_DECREF(ns);
+        if (!dummy_class)
+            goto error;
+
+        Py_DECREF(bases);
+        bases = dummy_class->tp_bases;
+        Py_INCREF(bases);
```

Comment:
I think these are basically necessary though I can drop some of the later refcount traffic on bases. We need to decref the bases here if we stop using it, assuming we got a new reference to it from `update_bases`, and we need to increment the refcount after we pull it out of the dummy class because we're going to stick it in some new object. We could wait until then to increment the refcount possibly, but that complicates error handling.

Some of the refcount traffic on bases that I am removing was basically to simplify error handling also: maintaining a reference to bases that we can decrement in error cases. It's easy to change that to `NULL` out the variable once the last reference is given away, though.